tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,230 @@
+    fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+        let s = format!("{:?}", value);
+        self.0.insert(field.name(), json!(s));
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: BigSpan) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, BigSpan>> {
+        self.spans.get(id_to_idx(id))
+    }
+
+    fn take(&self, id: Id) -> Option<BigSpan> {
+        self.spans.take(id_to_idx(&id))
+    }
+}
+
+thread_local! {
+    static CONTEXT: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let mut values = HashMap::new();
+        let mut visitor = RegistryVisitor(&mut values);
+        attrs.record(&mut visitor);
+        let parent = attrs.parent().map(|id| id.clone());
+        let s = BigSpan {
+            metadata: attrs.metadata(),
+            parent,
+            values: Mutex::new(values),
+        };
+        let id = (self.insert(s).expect("Unable to allocate another span") + 1) as u64;
+        Id::from_u64(id.try_into().unwrap())
+    }
+
+    #[inline]
+    fn record(&self, span: &span::Id, values: &span::Record<'_>) {
+        dbg!(span);
+        dbg!(values);
+    }
+
+    fn record_follows_from(&self, _span: &span::Id, _follows: &span::Id) {
+        // TODO: implement this please
+    }
+
+    fn enter(&self, id: &span::Id) {
+        let id = id.into_u64();
+        self.local_spans
+            .write()
+            .expect("Mutex poisoned")
+            .push(span::Id::from_u64(id));
+    }
+
+    fn event(&self, event: &Event<'_>) {
+        let id = match event.parent() {
+            Some(id) => Some(id.clone()),
+            None => {
+                if event.is_contextual() {
+                    self.local_spans
+                        .read()
+                        .expect("Mutex poisoned")
+                        .current()
+                        .map(|id| id.clone())
+                } else {
+                    None
+                }
+            }
+        };
+        if let Some(id) = id {
+            let mut values = HashMap::new();
+            let mut visitor = RegistryVisitor(&mut values);
+            event.record(&mut visitor);
+            let span = self.get(&id).expect("Missing parent span for event");
+        }
+    }
+
+    fn exit(&self, id: &span::Id) {
+        self.local_spans.write().expect("Mutex poisoned").pop(id);
+    }
+
+    #[inline]
+    fn try_close(&self, id: span::Id) -> bool {
+        let _ = self.take(id);
```

Comment:
can probably copy the span ref counting code from fmt, although perhaps it can be simplified